### PR TITLE
[GOBBLIN-2153]added search attributes to Temporal workflows

### PR DESCRIPTION
**Pull request by pratapaditya04**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2153

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Added flowGroup.flowName and user.to.proxy as  SearchAttributes to filter Temporal Flows in the UI

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+   * Generates search attributes for a WorkFlow  based on the provided GAAS job properties.
+   *
+   * @param jobProps the properties of the job, must not be null.
+   * @return a map containing the generated search attributes.
+   */
+  public static Map<String, Object> generateGaasSearchAttributes(@NonNull Properties jobProps) {
+    Map<String, Object> attributes = new HashMap<>();
+    attributes.put(Help.GAAS_FLOW_KEY, String.format("%s.%s", jobProps.getProperty(ConfigurationKeys.FLOW_GROUP_KEY),
+        jobProps.getProperty(ConfigurationKeys.FLOW_NAME_KEY)));
+    attributes.put(Help.USER_TO_PROXY_SEARCH_KEY, jobProps.getProperty(Help.USER_TO_PROXY_KEY));
```
Comment (khandelwal-prateek) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java:29]: let's add null checks for required keys in jobProps. If `FLOW_GROUP_KEY`, `FLOW_NAME_KEY`, or `USER_TO_PROXY_KEY` are missing, the method will currently throw a NullPointerException. It would be better to handle this gracefully by throwing an IllegalArgumentException with an error message.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+   *
+   * @param searchAttributes a map where the keys are attribute names and the values are lists of attribute values.
+   *                         Can be null.
+   * @return a map where the keys are attribute names and the values are the corresponding attribute values.
+   *         If the input map is null, an empty map is returned.
+   */
+  public static Map<String, Object> convertSearchAttributesValuesFromListToObject(
+      Map<String, List<?>> searchAttributes) {
+    if (searchAttributes == null) {
+      return null;
```
Comment (khandelwal-prateek) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java:44]: javadoc mentions that the method should return an empty map if searchAttributes is null, however, this returns null. We can use `Collections.emptyMap()` here instead of returning null

File: gobblin-temporal/src/test/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+public class TemporalWorkFlowUtilsTest {
+
+  private Properties jobProps = Mockito.mock(Properties.class);
+
+  @Test
+  public void testGenerateGaasSearchAttributes() {
+    // Mocking the properties
+    Mockito.when(jobProps.getProperty(ConfigurationKeys.FLOW_GROUP_KEY)).thenReturn("group");
+    Mockito.when(jobProps.getProperty(ConfigurationKeys.FLOW_NAME_KEY)).thenReturn("name");
+    Mockito.when(jobProps.getProperty(Help.USER_TO_PROXY_KEY)).thenReturn("userProxy");
```
Comment (khandelwal-prateek) [inline gobblin-temporal/src/test/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtilsTest.java:24]: the test assumes that the required properties(`FLOW_GROUP_KEY`, `FLOW_NAME_KEY` and `USER_TO_PROXY_KEY`) are always present. It would be useful to add a test where some of these properties are missing or null and test the behaviour in such cases

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    Map<String, Object> convertedAttributes = new HashMap<>();
+
+    convertedAttributes.putAll(searchAttributes);
```
Comment (khandelwal-prateek) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java:48]: It seems like the intention of this method is to convert the input map `Map<String, List<?>>` into a `Map<String, Object>` without transforming the list values into single objects. This method seems to be written for setSearchAttributes on ChildWorkflowOptions in AbstractNestingExecWorkflowImpl(which expects Map<String, Object>), we can simplify the implementation by directly using `new HashMap<>(searchAttributes)`. This would achieve the same result more efficiently.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 2 above)
Comment (Blazer-007) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java:44]: l believe using Optional<> can be more intuitive in such cases.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * Generates search attributes for a WorkFlow  based on the provided GAAS job properties.
+   *
+   * @param jobProps the properties of the job, must not be null.
+   * @return a map containing the generated search attributes.
+   */
+  public static Map<String, Object> generateGaasSearchAttributes(@NonNull Properties jobProps) {
+    Map<String, Object> attributes = new HashMap<>();
+    attributes.put(Help.GAAS_FLOW_KEY, String.format("%s.%s", jobProps.getProperty(ConfigurationKeys.FLOW_GROUP_KEY),
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java:27]: we should call this GAAS_FLOW_ID_SEARCH_KEY, as flowGroup + flowName is the UUID for a given flow config in GaaS

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java [hunk 6]
```diff
@@ -57,21 +59,25 @@ public CommitStats process(WUProcessingSpec workSpec) {
...
+      } else {
+        log.error("No work units processed, so no commit attempted.");
+        return CommitStats.createEmpty();
       }
-      return result;
-    } else {
-      log.error("No work units processed, so no commit attempted.");
+    } catch (Exception ignored) {
+      log.error("Exception occured during performing Work", ignored);
       return CommitStats.createEmpty();
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java:81]: What's the rationale for this try catch? Seems dangerous to ignore exceptions during processworkunits step, shouldn't we want it to fail loudly?

**Review by Will-Lo — CHANGES_REQUESTED** [review]

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 1 above)
Comment (pratapaditya04) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java]: have added default values so incase FLOW_GROUP_KEY, FLOW_NAME_KEY and USER_TO_PROXY_KEY are not set, the default values will be returned.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 2 above)
Comment (pratapaditya04) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java]: 1. addressed.
2. Do you mean Optional<> as the parameter to avoid null check or the return type?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 4 above)
Comment (pratapaditya04) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java]: simplified the conversion, however have kept this function as we need validations as new HashMap<>(searchAttributes)  throws NPE in case passed parameter is null

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 2 above)
Comment (Blazer-007) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java]: Optional<> as return type, since it is null safe by default.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java (hunk 6 above)
Comment (Will-Lo) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java]: If it throws an exception when loading the jobState (which I presume would lead to an empty jobState) we should just throw an ApplicationException in this scenario, so that we can push the Temporal framework to retry the workflow rather than progressing silently.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java (hunk 2 above)
Comment (pratapaditya04) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/util/TemporalWorkFlowUtils.java]: Don't think returning Optional is needed, as we are returning an empty Map instead of null.

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java (hunk 6 above)
Comment (pratapaditya04) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java]: makes sense, addressed

**Review by Will-Lo — COMMENTED** [review]
LGTM!
